feat: gravity acceleration

reginleiff opened the pull request:
w.r.t #34 

Currently breaks in quite a few cases:
- it applies universally to all movements
- block does not accelerate while falling after being moved left/right

Current implementation relies on`MoveBlock()` in `BlockBehaviour.cs`. My solution would require a duplicate version of `MoveBlock()` that handles falling separately but it may severely change the logic of move. Also, because of possible movements in left and right, I need to be able access the target block's bottom face which has been removed in a recent PR.

Also, I am unsure whether to put most of this implemented logic in `GravityPlugin`, because technically the logic still applies to the block behaviour. Perhaps I could store the constants in the plugin, but that would be very troublesome to experiment with as well.

Please advise @nicholaschuayunzhi

reginleiff committed:
feat: modify TranslateBlock to work for faster speeds

reginleiff committed:
feat: add acceleration to block behaviour

- change speed to reflect initial speed setting
- add acceleration factor
- add two methods to increase and reset speed

reginleiff committed:
feat: add terminal speed and change acceleration behaviour

nicholaschuayunzhi commented on Assets/Scripts/BlockBehaviour.cs:
```diff
@@ -86,29 +93,44 @@ public bool MoveBlock(BlockFace face)
         {
             _targetPosition = transform.position + (face.GetNormal() * _blockFaceBehaviour.GetFaceLength());
             _isTranslating = true;
+            _isFalling = true;
```
This isnt necessarily true? If a block moves on top of another block, it wouldnt be falling.

nicholaschuayunzhi commented on Assets/Scripts/BlockBehaviour.cs:
```diff
@@ -10,10 +10,15 @@ public class BlockBehaviour : MonoBehaviour
     public float SkinToLengthRatio = 0.1f;
-    public float Speed = 1f;
+
+    private float CurrentSpeed;
```
remember to follow our styling. You can download https://www.jetbrains.com/resharper/. its 👍

nicholaschuayunzhi replied:
If you haven't already done so.

nicholaschuayunzhi commented on Assets/Scripts/BlockBehaviour.cs:
```diff
@@ -86,29 +93,44 @@ public bool MoveBlock(BlockFace face)
+        if (_isFalling)
+        {
+            UpdateSpeed();
+        }
```
I think updating speed only when it is falling is too specific.

nicholaschuayunzhi commented on Assets/Scripts/BlockBehaviour.cs:
```diff
@@ -10,10 +10,15 @@ public class BlockBehaviour : MonoBehaviour
 
     private Vector3 _targetPosition;
 
+    private bool _isFalling;
```
Is it possible to move this into the gravity plugin?

nicholaschuayunzhi commented on Assets/Scripts/BlockBehaviour.cs:
```diff
@@ -10,10 +10,15 @@ public class BlockBehaviour : MonoBehaviour
+
+    private float CurrentSpeed;
+    public float InitialSpeed = 1f;
+    public float AccelerationFactor = 10f;
```
I would suggest using Acceleration instead of a factor. Speed Will be update as such:
```cs
CurrentSpeed += Acceleration * Time.deltaTime
```
To be done every frame.

nicholaschuayunzhi reviewed (changes requested):
What i suggest is to update the `MoveBlock(BlockFace)` to accept to more variables, initial speed and acceleration. And then setting `CurrentSpeed` to initial speed and using acceleration as the acceleration you want. That way, in the gravity plugin you can maintain the acceleration and speed and accelerate the block accordingly.

reginleiff committed:
feat: modify acceleration behaviour

- change MoveBlock to take in an additional parameter for acceleration
- shift acceleration constants to individual plugins
- update other plugins to reflect changes

nicholaschuayunzhi commented on Assets/Scripts/BlockBehaviour.cs:
```diff
@@ -79,10 +85,12 @@ private void Update()
+        Debug.Log(_currentSpeed);
         if (_isTranslating)
         {
+            UpdateSpeed(acceleration);
```
This check should not do anything but return false.

nicholaschuayunzhi replied:
MoveBlock should be called once to move a block by 1 unit.

nicholaschuayunzhi commented on Assets/Scripts/BlockBehaviour.cs:
```diff
@@ -95,20 +103,28 @@ public bool MoveBlock(BlockFace face)
-        transform.Translate(translateDir.normalized * Time.deltaTime * Speed);
-        if (Vector3.Distance(transform.position, _targetPosition) <= 0.01f)
+        Vector3 translate = translateDir.normalized * Time.deltaTime * _currentSpeed;
+
```
i suggest you store acceleration as a private variable and update speed here.

reginleiff committed:
feat: shift acceleration variables to BlockBehaviour

reginleiff committed:
feat: change updateSpeed to call per TranslateBlock instead of MoveBlock

nicholaschuayunzhi commented on Assets/ScriptableObjects/Plugin Scripts/MoveablePlugin.cs:
```diff
@@ -32,5 +31,7 @@ public override void OnFaceClick(BlockFace face)
             }
         }
+
+        _block.ResetSpeed();
```
I don't moveable should explicitly reset the speed.

nicholaschuayunzhi commented on Assets/ScriptableObjects/Plugin Scripts/MoveablePlugin.cs:
```diff
@@ -32,5 +31,7 @@ public override void OnFaceClick(BlockFace face)
             }
         }
+
+        _block.ResetSpeed();
```
I don't think move block should explicitly reset the speed.

nicholaschuayunzhi replied:
Instead could you have `MoveBlock` take in an initial speed that lets the plugin explicitly set what initial speed he wants. To make this even cleaner we can have a optional parameter for it. https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/named-and-optional-arguments

reginleiff committed:
feat: shift acceleration variables back to plugin

nicholaschuayunzhi committed:
feat: rewrite plugin to use MoveBlock only

nicholaschuayunzhi commented:
@reginleiff, I've updated the plugin as requested. Requesting for review. Also, sorry i had to do a force push because when i pushed the branch, github automatically merged which was not desired. As such my face appears next to yours.. 
![image](https://user-images.githubusercontent.com/22221132/37245450-fc20f994-24d2-11e8-9856-ead360b18076.png)
